pkg/controlplane/controller/leaderelection/leaderelection_controller.go

```diff
@@ -0,0 +1,428 @@
+)
+
+// Controller is the leader election controller, which observes component identity leases for
+// components that have self-nominated as candidate leaders for leases and elects leaders
+// for those leases, favoring candidates with higher versions.
+type Controller struct {
+	leaseInformer     coordinationv1informers.LeaseInformer
+	leaseClient       coordinationv1client.CoordinationV1Interface
+	leaseRegistration cache.ResourceEventHandlerRegistration
+
+	leaseCandidateInformer     coordinationv1alpha1.LeaseCandidateInformer
+	leaseCandidateClient       coordinationv1alpha1client.CoordinationV1alpha1Interface
+	leaseCandidateRegistration cache.ResourceEventHandlerRegistration
+
+	queue workqueue.TypedRateLimitingInterface[types.NamespacedName]
+}
+
+func (c *Controller) Run(ctx context.Context, workers int) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+	defer func() {
+		err := c.leaseInformer.Informer().RemoveEventHandler(c.leaseRegistration)
+		if err != nil {
+			klog.Warning("error removing leaseInformer eventhandler")
+		}
+		err = c.leaseCandidateInformer.Informer().RemoveEventHandler(c.leaseCandidateRegistration)
+		if err != nil {
+			klog.Warning("error removing leaseCandidateInformer eventhandler")
+		}
+	}()
+
+	if !cache.WaitForNamedCacheSync(controllerName, ctx.Done(), c.leaseRegistration.HasSynced, c.leaseCandidateRegistration.HasSynced) {
+		return
+	}
+
+	// This controller is leader elected and may start after informers have already started. List on startup.
+	lcs, err := c.leaseCandidateInformer.Lister().List(labels.Everything())
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+	for _, lc := range lcs {
+		c.enqueueCandidate(lc)
+	}
+
+	klog.Infof("Workers: %d", workers)
+	for i := 0; i < workers; i++ {
+		klog.Infof("Starting worker")
+		go wait.UntilWithContext(ctx, c.runElectionWorker, time.Second)
+	}
+	<-ctx.Done()
+}
+
+func NewController(leaseInformer coordinationv1informers.LeaseInformer, leaseCandidateInformer coordinationv1alpha1.LeaseCandidateInformer, leaseClient coordinationv1client.CoordinationV1Interface, leaseCandidateClient coordinationv1alpha1client.CoordinationV1alpha1Interface) (*Controller, error) {
+	c := &Controller{
+		leaseInformer:          leaseInformer,
+		leaseCandidateInformer: leaseCandidateInformer,
+		leaseClient:            leaseClient,
+		leaseCandidateClient:   leaseCandidateClient,
+
+		queue: workqueue.NewTypedRateLimitingQueueWithConfig(workqueue.DefaultTypedControllerRateLimiter[types.NamespacedName](), workqueue.TypedRateLimitingQueueConfig[types.NamespacedName]{Name: controllerName}),
+	}
+	leaseSynced, err := leaseInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueueLease(obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			c.enqueueLease(newObj)
+		},
+		DeleteFunc: func(oldObj interface{}) {
+			c.enqueueLease(oldObj)
+		},
+	})
+
+	if err != nil {
+		return nil, err
```

Comment:
non-blocker/cosmetics
```suggestion
	})
	if err != nil {
		return nil, err
```

Reply:
same below